modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java

```diff
@@ -3214,8 +3234,29 @@ private void body0() throws InterruptedException, IgniteCheckedException {
                         if (isCancelled())
                             break;
 
-                        if (task instanceof RebalanceReassignExchangeTask)
-                            exchId = ((RebalanceReassignExchangeTask) task).exchangeId();
+                        if (task instanceof RebalanceReassignExchangeTask) {
```

Comment:
You can use here last completed exchange future.
If it's version is higher, reassign task should be immediately cancelled. New exchange will trigger new rebalancing.

Reply:
We don't put a new reassign task into the exchange queue if it already contains a server exchange. Anyway, using the last completed exchange future with the condition looks reasonable to me. I will try to fix.